evap/staff/tools.py

```diff
@@ -68,31 +69,95 @@ def delete_navbar_cache_for_users(users):
+        elif matching_user.email != imported_email:
+            users_to_be_updated.append((matching_user, imported_email))
+            emails_of_non_obsolete_users.append(matching_user.email)  # add old address so this user won't be deleted
+
+    obsolete_users = UserProfile.objects.exclude(email__in=emails_of_non_obsolete_users)
+    deletable_users = [user for user in obsolete_users if user.can_be_deleted_by_manager]
+    users_to_mark_inactive = [
+        user for user in obsolete_users
+        if user not in deletable_users and user.is_active and user.can_be_marked_inactive_by_manager
+    ]
+
+    messages.info(
+        request,
+        _('The uploaded text file contains {} internal and {} external users. The external users will be ignored. '
+        '{} users are currently in the database. Of those, {} will be updated, {} will be deleted and {} will be '
+        'marked inactive. {} new users will be created.')
+        .format(len(imported_emails)-skipped_external_emails_counter, skipped_external_emails_counter,
+            UserProfile.objects.count(), len(users_to_be_updated), len(deletable_users), len(users_to_mark_inactive),
+            len(emails_of_users_to_be_created))
+    )
+    if users_to_be_updated:
+        messages.info(
+            request,
+            mark_safe(
+                _('Users to be updated are:<br />{}')
+                .format('<br />'.join(
+                    [_("{} > {}").format(user.email, email) for user, email in users_to_be_updated]
+                ))
+            )
+        )
+    if deletable_users:
+        messages.info(request, mark_safe(_('Users to be deleted are:<br />{}')
+            .format('<br />'.join([user.email or user.username for user in deletable_users]))))
+    if users_to_mark_inactive:
+        messages.info(request, mark_safe(_('Users to be marked inactive are:<br />{}')
+            .format('<br />'.join([user.email or user.username for user in users_to_mark_inactive]))))
+    if emails_of_users_to_be_created:
+        messages.info(request, mark_safe(_('Users to be created are:<br />{}')
+            .format('<br />'.join(emails_of_users_to_be_created))))
 
     if test_run:
-        messages.info(request, _('No users were deleted or marked inactive in this test run.'))
+        messages.info(request, _('No data was changed in this test run.'))
     else:
         for user in deletable_users:
             user.delete()
         for user in users_to_mark_inactive:
             user.is_active = False
             user.save()
-
-        messages.info(request, _('{} users have been deleted').format(len(deletable_users)))
-        messages.info(request, _('{} users have been marked inactive').format(len(users_to_mark_inactive)))
+        for user, email in users_to_be_updated:
+            user.email = email
+            user.save()
+        for email in emails_of_users_to_be_created:
+            UserProfile.objects.create(email=email, username=email)
```

Comment:
You wanted to use a transaction here ^^

Reply:
Or use [bulk_create](https://docs.djangoproject.com/en/3.0/ref/models/querysets/#bulk-create)?